src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -222,51 +223,242 @@ private bool ScanDesignation(bool permitTuple)
                     {
-                        // X.Y.Z id
-                        if (this.IsTrueIdentifier() && this.CurrentToken.ContextualKind != SyntaxKind.WhenKeyword)
-                        {
-                            var designation = ParseSimpleDesignation();
-                            node = _syntaxFactory.DeclarationPattern(type, designation);
-                        }
+                        // either it is not shaped like a type, or it is a constant expression.
+                        this.Reset(ref resetPoint);
+                        type = null;
+                    }
+                }
+
+                PropertySubpatternSyntax propertySubpattern = null;
+                bool parsePropertySubpattern()
+                {
+                    if (this.CurrentToken.Kind == SyntaxKind.OpenBraceToken)
+                    {
+                        propertySubpattern = ParsePropertySubpattern();
+                        return true;
+                    }
+
+                    return false;
+                }
+
+                VariableDesignationSyntax designation = null;
+                bool parseDesignation()
+                {
+                    if (this.IsTrueIdentifier() && (!forCase || this.CurrentToken.ContextualKind != SyntaxKind.WhenKeyword))
+                    {
+                        designation = ParseSimpleDesignation();
+                        return true;
+                    }
+
+                    return false;
+                }
+
+                bool looksLikeCast()
+                {
+                    var resetPoint2 = this.GetResetPoint();
+                    try
+                    {
+                        return this.ScanCast(forPattern: true);
                     }
-                    if (node == null)
+                    finally
                     {
-                        // it is an expression for typical switch case. 
+                        this.Reset(ref resetPoint2);
+                        this.Release(ref resetPoint2);
+                    }
+                }
+
+                if (this.CurrentToken.Kind == SyntaxKind.OpenParenToken && (type != null || !looksLikeCast()))
+                {
+                    // It is possible this is a parenthesized (constant) expression.
+                    // We normalize later.
+                    ParseSubpatternList(out var openParenToken, out var subPatterns, out var closeParenToken, SyntaxKind.OpenParenToken, SyntaxKind.CloseParenToken);
+                    if (this.CurrentToken.Kind == SyntaxKind.EqualsGreaterThanToken)
+                    {
+                        // Testers do the darndest things, in this case putting a lambda expression where a pattern is expected.
                         this.Reset(ref resetPoint);
-                        node = this.ParseSubExpression(Precedence.Expression);
+                        return this.ParseSubExpression(Precedence.Expression);
```

Comment:
>return this.ParseSubExpression(Precedence.Expression); [](start = 24, length = 54)

If we have a ```type (...) =>```, why is this a good thing to reparse as expression?
Even if we don't have a type, but have a sub-pattern  in the form of ```identifier : ... ```, this still doesn't look like a lambda.

Reply:
I found that this improved the quality of our syntax error recovery in a number of existing (intentionally erroneous) syntax tests.

---
In reply to: [151813444](https://github.com/dotnet/roslyn/pull/23208#discussion_r151813444) [](ancestors = 151813444)

Reply:
I didn't notice any existing tests affected by this PR that cover scenarios like that. I also didn't notice any new tests covering scenarios like that. I would agree that for scenarios where at least lambda's signature would parse successfully, this recovery makes sense. However, this logic covers a lot more other scenarios. So, I believe the logic requires thorough testing and quite possibly refingment.

---
In reply to: [151814567](https://github.com/dotnet/roslyn/pull/23208#discussion_r151814567) [](ancestors = 151814567,151813444)

Reply:
I don't know what you mean when you say "this logic covers a lot more other scenarios". This method is only used to parse where a pattern is expected, within which a lambda is never legal. An error in such contexts (e.g. a case expression) appears only in our test code to test error recovery. Parsing the case as an expression (possibly a lambda expression) ensures consistency with previous error recovery, where the case body was parsed as an expression. That is why you didn't see tests modified to account for this code; this code is present to accommodate the tests. I'm fine with additional testing and refinement of the error recovery.

---
In reply to: [151820288](https://github.com/dotnet/roslyn/pull/23208#discussion_r151820288) [](ancestors = 151820288,151814567,151813444)